pkg/controller/drain.go

```diff
@@ -487,85 +493,112 @@ func filterSharedPVs(pvMap map[string][]string) {
 				returnCh <- nil
-				continue
-			} else if err != nil {
-				returnCh <- fmt.Errorf("Error when evicting pod: %v/%v. Will not be retried. Err: %v", pod.Namespace, pod.Name, err)
-				continue
 			}
+		}
 
-			glog.V(3).Info("waiting for PVs to detach from node for pod: ", pod.Name)
-			// Eviction was successful. Wait for pvs for this pod to detach
-			pvs := volMap[pod.Namespace+"/"+pod.Name]
-			err = o.waitForDetach(pvs, o.nodeName, stopCh)
+		if len(remainingPods) > 0 {
+			// Force delete the pods remaining after evict retries. No pods are expected to survive this and errors
+			// are returned via returnCh anyway.
+			_, _ = o.evictPodsWithPVInternal(false, pods, volMap, policyGroupVersion, getPodFn, returnCh)
+		}
+	} else {
+		// Force delete the pods remaining after evict retries. No pods are expected to survive this and errors
+		// are returned via returnCh anyway.
+		_, _ = o.evictPodsWithPVInternal(false, pods, volMap, policyGroupVersion, getPodFn, returnCh)
+	}
 
-			if apierrors.IsNotFound(err) {
-				glog.V(3).Info("Node not found anymore")
-				returnCh <- nil
-				remainingPods = len(pods) - (i + 1) + len(retryPods)
-				// The node itself is gone. Stop evicting pods.
-				// We should anyway return successful eviction for the remaining pods
-				break EvictingPods
-			} else if err != nil {
-				glog.Errorf("Error when waiting for volume to detach from node. Err: %v", err)
-				returnCh <- err
-				continue
-			}
-			glog.V(3).Info("Detached pv for pod: ", pod.Name)
+	return
+}
+
+func (o *DrainOptions) evictPodsWithPVInternal(attemptEvict bool, pods []*corev1.Pod, volMap map[string][]string,
+	policyGroupVersion string,
+	getPodFn func(namespace, name string) (*api.Pod, error),
+	returnCh chan error) (remainingPods []*api.Pod, fastTrack bool) {
+	var retryPods []*api.Pod
+	for i, pod := range pods {
+		var err error
+		if attemptEvict {
+			err = o.evictPod(pod, policyGroupVersion)
+		} else {
+			err = o.deletePod(pod)
+		}
+
+		if apierrors.IsTooManyRequests(err) {
+			// Pod eviction failed because of PDB violation, we will retry one we are done with this list.
+			glog.V(3).Info("Pod ", pod.Namespace, "/", pod.Name, " couldn't be evicted because of PDB violation. Will be retried.")
+			retryPods = append(retryPods, pod)
+			continue
+		} else if apierrors.IsNotFound(err) {
+			glog.V(3).Info("\t", pod.Name, " is already gone")
 			returnCh <- nil
+			continue
+		} else if err != nil {
+			glog.V(4).Infof("Error when evicting pod: %v/%v. Will be retried. Err: %v", pod.Namespace, pod.Name, err)
+			retryPods = append(retryPods, pod)
+			continue
 		}
 
-		if len(retryPods) == 0 {
-			// There are no pods to retry
-			break
+		// Eviction was successful. Wait for pvs for this pod to detach
+		glog.V(3).Info("Eviction/delete succeeded. Waiting for the pod to terminate: ", pod.Name)
+		_, err = o.waitForDelete([]api.Pod{*pod}, Interval, o.getTerminationGracePeriod(pod), true, getPodFn)
```

Comment:
Because of the multi-container case.

Reply:
Also, it is possible that the terminationGracePeriod is longer than the pvDetachTimeout.

Reply:
In that case we can do `waitForDelete` after we have triggered eviction for all serially. This would make sure that the detach is serial, and also that all the pods are terminated.